File: Problems/boj_섬의개수.py
#https://www.acmicpc.net/problem/4963
#DFS/BFS

from collections import deque
import sys
input = sys.stdin.readline

moves = ((-1,0), (1,0),(0,-1),(0,1),(-1,-1),(-1,1),(1,1),(1,-1))
def bfs(m, s, v):
    global w, h
    q = deque([s])
    v[s[0]][s[1]] = True

    while q:
        x, y = q.popleft()
        for move in moves:
            nx = x + move[0]
            ny = y + move[1]

            if not 0<=nx<h or not 0<=ny<w:
                continue
                
            if m[nx][ny]==1 and v[nx][ny]==False:
                v[nx][ny] = True
                q.append((nx, ny))
    
while True:
    w, h = map(int, input().split())
    if w==0 and h==0: break

    mat = [list(map(int, input().split())) for _ in range(h)]
    visited = [[False]*w for _ in range(h)]

    result = 0
    for i in range(h):
        for j in range(w):
            if mat[i][j]==1 and visited[i][j]==False:
                bfs(mat, (i,j), visited)
                result += 1
    print(result)

File: Problems/cpp/boj_섬의개수.cpp
//https://www.acmicpc.net/problem/4963
//BFS, DFS

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int dx[] = {1, -1, 0, 0, 1, 1, -1, -1};
int dy[] = {0, 0, -1, 1, 1, -1, 1, -1};

int r, c;
vector<vector<int>> mat;
vector<vector<bool>> visit;

void bfs(int i, int j){
    deque<pair<int, int>> dq;
    dq.push_back(make_pair(i, j));
    visit[i][j] = true;

    while (!dq.empty()){
        int x = dq.front().first;
        int y = dq.front().second;
        dq.pop_front();

        for (int i=0; i<8; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (0>nx || 0>ny || r<=nx || c<=ny) continue;

            if (mat[nx][ny]==1 && visit[nx][ny]==false){
                visit[nx][ny] = true;
                dq.push_back(make_pair(nx, ny));
            }
        } 
    }
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    vector<int> result;

    while (true){
        cin >> c >> r;
        
        if (c==0 && r==0){
            for (const int r : result)
                cout << r << '\n';
            break;
        }

        mat = vector<vector<int>>(r, vector<int>(c));
        for (int i=0; i<r; i++){
            for (int j=0; j<c; j++)
                cin >> mat[i][j];
        }

        int count=0;
        visit = vector<vector<bool>>(r, vector<bool>(c, false));
        
        for (int i=0; i<r; i++){
            for (int j=0; j<c; j++){
                if (mat[i][j]==1 && !visit[i][j]){
                    count++;
                    bfs(i, j);
                }
            }
        }
        
        result.push_back(count);
    }
}